### PR TITLE
Hijack IsLatest(Stable)Version OData filter when semVerLevel=2.0.0

### LINKED ISSUE #3905: $filter=IsLatestVersion and semVerLevel=2.0.0 yields no results
The client does the following query when using the Browse tab is VS:
```http
GET https://dev.nugettest.org/api/v2/Search()?$filter=IsLatestVersion&searchTerm=''&targetFramework='net46'&includePrerelease=false&$skip=0&$top=26&semVerLevel=2.0.0 HTTP/1.1
Accept: application/atom+xml, application/xml
user-agent: NuGet Client V3/4.3.0 (Microsoft Windows NT 10.0.14393.0, VS Enterprise/15.0)
X-NuGet-Client-Version: 4.3.0
Accept-Language: en-US
Host: dev.nugettest.org
Accept-Encoding: gzip, deflate
```

This yields no results. I think perhaps its a bug with the introduction of `IsLatestSemVer2` and `IsLatestStableSemVer2`.

This results in an empty "browse" tab when using SemVer 2.0.0:
image

### PR DESCRIPTION
**Pull request by xavierdecoster**
Fixes #3905

### CODE REVIEW

File: src/NuGetGallery/WebApi/QueryResult.cs [hunk 1]
```diff
@@ -1,19 +1,20 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.Data.OData;
+using Microsoft.Data.OData.Query;
```
Comment (chenriksson) [inline src/NuGetGallery/WebApi/QueryResult.cs]: nit: move under System namespaces

File: src/NuGetGallery/WebApi/QueryResult.cs [hunk 2]
```diff
@@ -121,7 +126,22 @@ public IHttpActionResult GetInnerResult()
 
                 if (queryOptions.Filter != null)
                 {
-                    queryResults = queryOptions.Filter.ApplyTo(queryResults, _querySettings);
+                    if (_semVerLevelKey != SemVerLevelKey.Unknown
+                        && (string.Equals(queryOptions.Filter.RawValue, "IsLatestVersion", StringComparison.OrdinalIgnoreCase)
+                            || string.Equals(queryOptions.Filter.RawValue, "IsAbsoluteLatestVersion", StringComparison.OrdinalIgnoreCase)))
```
Comment (chenriksson) [inline src/NuGetGallery/WebApi/QueryResult.cs]: Add constants for these? Looks like these strings are used in SearchAdapter as well.

File: tests/NuGetGallery.Facts/Controllers/ODataV2CuratedFeedControllerFacts.cs [hunk 3]
```diff
@@ -115,52 +115,132 @@ public class ODataV2CuratedFeedControllerFacts
...
+            // Assert
+            Assert.Equal(NonSemVer2Packages.Where(p => !p.IsPrerelease).Count(), searchCount);
+        }
+
+        [Theory]
+        [InlineData("2.0.0")]
+        [InlineData("2.0.1")]
+        [InlineData("3.0.0-alpha")]
+        [InlineData("3.0.0")]
+        public async Task SearchCount_IncludesSemVerV2PackageVersionsWhenSemVerLevel2OrHigher_IncludePrerelease(string semVerLevel)
```
Comment (chenriksson) [inline tests/NuGetGallery.Facts/Controllers/ODataV2CuratedFeedControllerFacts.cs]: Is there a test for semverlevel < 2 to ensure it resolves as unknown?

File: src/NuGetGallery/WebApi/QueryResult.cs [hunk 4]
```diff
@@ -59,6 +61,9 @@ public QueryResult(ODataQueryOptions<TModel> queryOptions, IQueryable<TModel> qu
             _totalResults = totalResults;
             _generateNextLink = generateNextLink;
 
+            var queryDictionary = HttpUtility.ParseQueryString(queryOptions.Request.RequestUri.Query);
+            _semVerLevelKey = SemVerLevelKey.ForSemVerLevel(queryDictionary["semVerLevel"]);
```
Comment (ryuyu) [inline src/NuGetGallery/WebApi/QueryResult.cs]: Is this ok if semVerLevel isn't passed? Or are we guaranteed to always pass semVerLevel? Should we do a TryGet instead?

File: tests/NuGetGallery.Facts/Controllers/ODataV2CuratedFeedControllerFacts.cs (hunk 3 above)
Comment (xavierdecoster) [inline tests/NuGetGallery.Facts/Controllers/ODataV2CuratedFeedControllerFacts.cs]: Yes, there's `SearchCount_FiltersSemVerV2PackageVersions_IncludePrerelease` and `SearchCount_FiltersSemVerV2PackageVersions_ExcludePrerelease`

File: src/NuGetGallery/WebApi/QueryResult.cs (hunk 4 above)
Comment (xavierdecoster) [inline src/NuGetGallery/WebApi/QueryResult.cs]: This is OK with `semVerLevel=null` being passed :) It uses a [`NameValueCollection`](https://msdn.microsoft.com/en-us/library/8d0bzeeb.aspx) under the hood, for which the index property returns `null` if the key is not found (which in turn is handled in the `SemVerLevelKey.ForSemVerLevel`

File: src/NuGetGallery/WebApi/QueryResult.cs (hunk 2 above)
Comment (xavierdecoster) [inline src/NuGetGallery/WebApi/QueryResult.cs]: Fixed in https://github.com/NuGet/NuGetGallery/pull/3966/commits/48846ad34658ae0af0ddea292849cb033305209d

File: src/NuGetGallery/WebApi/QueryResult.cs (hunk 1 above)
Comment (xavierdecoster) [inline src/NuGetGallery/WebApi/QueryResult.cs]: Fixed in https://github.com/NuGet/NuGetGallery/pull/3966/commits/48846ad34658ae0af0ddea292849cb033305209d

**Pull request merged**